### PR TITLE
Add disttae's logic of engine interface and database interface

### PR DESCRIPTION
**Pull request by nnsgmsone**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3931

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/vm/engine/disttae/database.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// limitations under the License.
+
+package disttae
+
+import (
+	"context"
+
+	"github.com/matrixorigin/matrixone/pkg/vm/engine"
+)
+
```
Comment (ouyuanning) [inline pkg/vm/engine/disttae/database.go:22]: can add : var _ engine.Database = new(database)

File: pkg/vm/engine/disttae/table.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import (
+	"context"
+	"strconv"
+
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine"
+)
+
```
Comment (ouyuanning) [inline pkg/vm/engine/disttae/table.go:26]: can add :   var _ engine.Relation = new(table)

File: pkg/vm/engine/disttae/types.go [hunk 3]
```diff
@@ -47,6 +47,44 @@ const (
...
 
+// column's index in catalog table
+const (
+	MO_DATABASE_DAT_ID_IDX   = 0
+	MO_DATABASE_DAT_NAME_IDX = 1
+	MO_TABLES_REL_ID_IDX     = 0
+	MO_TABLES_REL_NAME_IDX   = 1
+)
+
+var (
```
Comment (reusee) [inline pkg/vm/engine/disttae/types.go:58]: These names should be generated from catalog.SystemDBSchema, SystemTableSchema, SystemColumnSchema

File: pkg/vm/engine/disttae/types.go [hunk 4]
```diff
@@ -47,6 +47,44 @@ const (
...
+		"dat_id",
+		"datname",
+		"dat_catalog_name",
+		"dat_createsql",
+		"owner",
+		"creator",
+		"created_time",
+		"account_id",
+	}
+	MoTablesSchema = []string{
```
Comment (ouyuanning) [inline pkg/vm/engine/disttae/types.go:69]: I don't known if these columns will contains all of the tableDef（eg:  ViewDef,  CommentDef, IndexDef,  ExternalTable)

File: pkg/vm/engine/disttae/types.go (hunk 3 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/types.go:58]: ok

File: pkg/vm/engine/disttae/types.go (hunk 4 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/types.go:69]: yes

**Review by ouyuanning — APPROVED** [review]

File: pkg/catalog/types.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+package catalog
+
+const (
+	// default database name for catalog
+	MO_CATALOG  = "mo_catalog"
+	MO_DATABASE = "mo_database"
+	MO_TABLES   = "mo_tables"
+	MO_COLUMNS  = "mo_columns"
+)
```
Comment (reusee) [inline pkg/catalog/types.go:23]: These should not be string constants but reference to these:
https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L43
https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L45
https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L46
https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L47

File: pkg/catalog/types.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+const (
+	// default database id for catalog
+	MO_CATALOG_ID  = 1
+	MO_DATABASE_ID = 1
+	MO_TABLES_ID   = 2
+	MO_COLUMNS_ID  = 3
+)
```
Comment (reusee) [inline pkg/catalog/types.go:31]: These constants are also define in tae catalog packages:
https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L48

File: pkg/catalog/types.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		"relpersistence",
+		"relkind",
+		"rel_comment",
+		"rel_createsql",
+		"created_time",
+		"creator",
+		"owner",
+		"account_id",
+	}
+	MoColumnsSchema = []string{}
```
Comment (reusee) [inline pkg/catalog/types.go:66]: These strings should be constructed by looping SystemDBSchema.ColDefs

https://github.com/matrixorigin/matrixone/blob/263a69919a1bd73ade7846aa2461bcd01b590c10/pkg/vm/engine/tae/catalog/model.go#L118

**Review by reusee — APPROVED** [review]

**Comment by nnsgmsone** [comment]
https://github.com/matrixorigin/docs/blob/main/cn/cn.md，This is the documentation for cn, see also.

File: pkg/vm/engine/disttae/database.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	return &table{
+		tableId:   id,
+		tableName: name,
+		db:        db,
+	}, nil
+}
+
+func (db *database) Delete(ctx context.Context, name string) error {
+	if err := db.txn.WriteBatch(DELETE, catalog.MO_CATALOG_ID, catalog.MO_TABLES_ID,
```
Comment (fengttt) [inline pkg/vm/engine/disttae/database.go:43]: Why is this WriteBatch?   Seems we are writing one by one?

File: pkg/vm/engine/disttae/table.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+func (tbl *table) AddTableDef(ctx context.Context, def engine.TableDef) error {
+	return nil
+}
+
+func (tbl *table) DelTableDef(ctx context.Context, def engine.TableDef) error {
+	return nil
+}
+
+func (tbl *table) GetTableID(ctx context.Context) string {
```
Comment (fengttt) [inline pkg/vm/engine/disttae/table.go:77]: why table id is string?   why not just tbl.tableId?

File: pkg/vm/engine/disttae/table.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+func (tbl *table) Rows() int64 {
+	return 0
+}
+
+func (tbl *table) Size(name string) int64 {
+	return 0
+}
+
+func (tbl *table) Ranges(ctx context.Context) ([][]byte, error) {
```
Comment (fengttt) [inline pkg/vm/engine/disttae/table.go:37]: What is Ranges?

File: pkg/vm/engine/disttae/table.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+func (tbl *table) TableDefs(ctx context.Context) ([]engine.TableDef, error) {
+	return nil, nil
+}
+
+func (tbl *table) GetPrimaryKeys(ctx context.Context) ([]*engine.Attribute, error) {
+	return nil, nil
+}
+
+func (tbl *table) GetHideKeys(ctx context.Context) ([]*engine.Attribute, error) {
```
Comment (fengttt) [inline pkg/vm/engine/disttae/table.go:49]: HiddenKeys

File: pkg/vm/engine/disttae/table.go (hunk 9 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/table.go:77]: > why table id is string? why not just tbl.tableId?

I don't know why. This is 0.5 tae so designed, I keep this for now.

File: pkg/vm/engine/disttae/table.go (hunk 11 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/table.go:49]: > HiddenKeys


I don't even think this interface should be needed, with the TableDefs() interface you can fetch all the information of the table. This type of interface my plan is to remove after 0.6 is ready to run.

File: pkg/vm/engine/disttae/database.go (hunk 8 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/database.go:43]: Yes, but the mergeentry function will be added later.

File: pkg/vm/engine/disttae/table.go (hunk 10 above)
Comment (nnsgmsone) [reply pkg/vm/engine/disttae/table.go:37]: > What is Ranges?


For tae it is the block list, for memengine it is the range of primary keys, like [0 - 10), [10, 20), [20, 30), [30, max].
This interface is mainly used to slice and dice queries into different pipelines. @fengttt

**Comment by nnsgmsone** [comment]
The engine interface now has many things that I think are strange, and I will try to modify these interfaces as well as the definitions in 0.6.

**Review by fengttt — APPROVED** [review]

**Pull request merged**